### PR TITLE
refactored reodering of orbitals

### PR DESCRIPTION
**Pull request by rubengerritsen**
Just the refactoring of the orbital reordering

### CODE REVIEW

**Comment by rubengerritsen** [comment]
@votca-bot changelog: Refactor orbital reordering

**Comment by rubengerritsen** [comment]
@votca-bot format

File: include/votca/xtp/qmatom.h [hunk 1]
```diff
@@ -65,6 +65,8 @@ class QMAtom {
 
   Index getId() const { return _index; }
 
+  Index getPureNucCharge() const { return _nuccharge; }
```
[the remark this answers is not in the document]
Comment (rubengerritsen) [reply include/votca/xtp/qmatom.h:68]: Stems from the molden parser, I'll move it to that PR

File: src/libxtp/orbreorder.cc [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+// Local VOTCA includes
+#include "votca/xtp/orbreorder.h"
+
+namespace votca {
+namespace xtp {
+
+void OrbReorder::reorderOrbitals(Eigen::MatrixXd& moCoefficients,
+                                 AOBasis& basis) {
```
[the remark this answers is not in the document]
Comment (rubengerritsen) [reply src/libxtp/orbreorder.cc:27]: Yes it can

File: src/libxtp/orbreorder.cc [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+  return std::vector<Index>{input.begin() + start,
+                            input.begin() + start + size};
+}
+
+OrbReorder::OrbReorder(std::array<Index, 25> reorder,
+                       std::array<Index, 25> multipliers, bool reverse)
+    : _multipliers(multipliers), _reorder(reorder) {
+
+  // Compute transpositions for every individual shell
+  int currentFunction = 0;
```
Comment (JensWehner) [inline src/libxtp/orbreorder.cc:57]: could you use `Index` instead of `int` it makes refactoring easier.

File: src/libxtp/orbreorder.cc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  for (int l = 0; l < 5; l++) {
+    int nrOfFunctions = 2 * l + 1;
```
Comment (JensWehner) [inline src/libxtp/orbreorder.cc:59]: if you could use the the function from `basisset.h` to make it more explicit

File: src/libxtp/orbreorder.cc [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                                 const AOBasis& basis) {
+  std::vector<Index> multiplier;
+  multiplier.reserve(basis.AOBasisSize());
+
+  Index currentFunction = 0;
+  for (const AOShell& shell : basis) {
+
+    // reorder shell
+    Index l = static_cast<Index>(shell.getL());
+    for (auto& transposition : _transpositions[l]) {
```
Comment (JensWehner) [inline src/libxtp/orbreorder.cc:83]:
```suggestion
    for (const Transposition& transposition : _transpositions[l]) {
```

File: src/libxtp/orbreorder.cc [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      moCoefficients.row(currentFunction + std::get<0>(transposition))
```
Comment (JensWehner) [inline src/libxtp/orbreorder.cc:84]:
```suggestion
      moCoefficients.row(currentFunction + transposition.first)
```

File: src/libxtp/orbreorder.cc [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+          .swap(
+              moCoefficients.row(currentFunction + std::get<1>(transposition)));
```
Comment (JensWehner) [inline src/libxtp/orbreorder.cc:86]:
```suggestion
              moCoefficients.row(currentFunction + transposition.second));
```

File: src/libxtp/orbreorder.cc [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+    // Get multiplier vector for shell
+    std::vector<Index> shellmultiplier =
+        copySegment(_multipliers, shell.getOffset(), shell.getNumFunc());
+
+    // multiply shell
+    for (Index i = 0; i < shell.getNumFunc(); i++) {
+      moCoefficients.row(currentFunction + i) *= double(shellmultiplier[i]);
+      ;
+    }
```
Comment (JensWehner) [inline src/libxtp/orbreorder.cc:97]:
```suggestion
    for (Index i = 0; i < shell.getNumFunc(); i++) {
      moCoefficients.row(currentFunction + i) *= double(shellmultiplier[i]);
    }
```

**Review by JensWehner — APPROVED** [review]
@rubengerritsen I like the changes a lot. It cleans up qmpackage and makes the ordering much clearer Well done . I approve it already but please add the small changes still

**Comment by JensWehner** [comment]
@junghans  did you update our buildenv in the last week? I do not know where the openmp error is coming from. It was not there before. So it is a bit subtle.

**Comment by junghans** [comment]
> @junghans did you update our buildenv in the last week? I do not know where the openmp error is coming from. It was not there before. So it is a bit subtle.

The buildenv automatically get rebuild every week.

**Comment by junghans** [comment]
The error seems related to the new clang-11.

**Comment by JensWehner** [comment]
yeah that was my guess as well.

**Comment by JensWehner** [comment]
what is clang-11 ? I thought it is not even out yet. I cannot find any documentation on it? It is fedora specific?

**Comment by junghans** [comment]
> what is clang-11 ? I thought it is not even out yet. I cannot find any documentation on it? It is fedora specific?

Rawhide always has the latest and greatest compilers. clang-11 is a pre-release.

**Pull request merged**